CORE-1165 Integrate Metadata listing in data/ds/item page

psarando opened the pull request:
This PR will integrate the Metadata listing with `/data/ds/[...pathItem]` pages.

A new `?view=metadata` navigation query param constant was added that this page can use to determine whether to display metadata on a file or folder, rather than that item's contents.

For example, `/data/ds/iplant/home/ipcdev/bilbo-funny-1.jpg?view=metadata`.
![bilbo-funny-1.jpg Metadata Listing](https://user-images.githubusercontent.com/996408/102160611-fee33980-3e42-11eb-9d51-9927c2842f9d.png)

I also reorganized the data listing dot menus a bit, especially since it was far too easy for me to accidentally click `Delete` right after opening the dot menu when it was the first item in the list. So now `Delete` is at the bottom under a separating divider:
![Data Listing Toolbar Dot-menu](https://user-images.githubusercontent.com/996408/102160819-64cfc100-3e43-11eb-91ab-7bcf53606d54.png)

---

On mobile:
![Data Listing Toolbar Dot-menu - mobile](https://user-images.githubusercontent.com/996408/102161099-ecb5cb00-3e43-11eb-92a2-54e7ec43a970.png)

---

Similarly the listing row dot-menus were rearranged:
![Data Listing Row Dot-menu - mobile](https://user-images.githubusercontent.com/996408/102161217-22f34a80-3e44-11eb-9980-1d00c9bba97b.png)

psarando commented on src/components/metadata/form/index.js:
```diff
@@ -440,11 +446,19 @@ const MetadataForm = ({ loading, showErrorAnnouncer, ...props }) => {
             <Formik
                 enableReinitialize
-                initialValues={{ ...metadata }}
+                initialValues={{ avus, "irods-avus": irodsAVUs }}
```
I guess the DE's GXT code was sanitizing the service response before passing it to the old React code, since the metadata endpoint also returns a `path` in its response that is not acceptable in the metadata add/update endpoints. So this form is now explicitly only using the keys required in the form and submission.

psarando commented on src/common/NavigationConstants.js:
```diff
@@ -1,4 +1,7 @@
-// constants that define navigation routes
+/**
+ * Constants that define navigation routes.
+ * Warning: server/index.js expects each value to be a string.
```
Initially I tried adding the `VIEW: { METADATA: "metadata" }` object to these default constants, but I got a `string.replace is not a function` error message (without a stack trace) that I eventually tracked down to `server/index.js`.

So I added this comment to prevent others from running into the same issue 😅

sri2k1us commented:
LGTM. However, when I ran this locally I found couple UX improvements that could be made:
1. Move the `Add` avu to toolbar and call it out in large screen instead of just icon. It took me few seconds to figure out where the add button was.
![image](https://user-images.githubusercontent.com/1250790/102234542-64750b80-3ebf-11eb-995e-035038e7c8ee.png)

2. A `Back` button to take the user back to the listing.

3. I prefer the `Delete` button to be colored rather than its hover background color getting changed to red. In other places too, we set the color for icon (for example to `primary`) and not change the hover background color. Also in mobile, there is no hover rather the color shows up when you actually press the button. 
![image](https://user-images.githubusercontent.com/1250790/102235729-a3f02780-3ec0-11eb-8d7f-d83683e6ffdd.png)

4. May be `Metadata` action could be linked into the file viewer too ? (This could be saved for later though)

ash3rz commented on src/components/data/toolbar/DataDotMenu.js:
```diff
@@ -259,6 +260,15 @@ function DataDotMenu(props) {
                                 onClose={onClose}
                             />
                         ),
+                    <Divider key={ids.DELETE_MENU_ITEM_DIVIDER} />,
```
👍

ash3rz reviewed (approved):
LGTM :shipit: 
I mostly agree with Sriram's UX comments as well, though I wanted to add that I think the Add AVU button should remain visible on mobile if possible since that's probably a primary action.  Feels weird if that ends up getting hidden within the dot menu on mobile, to me at least.

psarando commented:
> LGTM. However, when I ran this locally I found couple UX improvements that could be made:

Most of this UX is a direct import from the existing UX in the DE, and I was hoping UX changes for Sonora could have been brought up in #238, and to keep this PR just for page integration and URL / listing navigation UX.

FYI, if you don't want to checkout a PR and spin up Storybook locally, Chromium will still build our stories, even when our snapshot limit is reached: https://5ebad5c6b991d60022b1e72e-hutdrspubj.chromatic.com/?path=/story/metadata--metadata-view

With that said, I think I can implement some of these improvements in another PR, but let me address them 1 at a time, below:

> 1. Move the `Add` avu to toolbar and call it out in large screen instead of just icon. It took me few seconds to figure out where the add button was. Suggested change:

This is a good idea, however, the `+ AVU` header and the listing below it are reused for nested/child metadata listings:
![Edit AVU Dialog](https://user-images.githubusercontent.com/996408/102406378-b4c79880-3fa7-11eb-9558-93af04926c2b.png)

So I'm not sure where to move this button in the AVU Dialog 🤔 

> 2. A `Back` button to take the user back to the listing.

This shouldn't be a problem and I can add this in this PR (this is the kind of Sonora UX navigation feedback I was looking for in this PR) 👍 

> 3. I prefer the `Delete` button to be colored rather than its hover background color getting changed to red. In other places too, we set the color for icon (for example to `primary`) and not change the hover background color. Also in mobile, there is no hover rather the color shows up when you actually press the button.

I agree, the delete icon should probably be solid red, instead of red on hover, which will be nicer for mobile users.
I'll save this for a UX improvement PR, though.

> 4. May be `Metadata` action could be linked into the file viewer too ? (This could be saved for later though)

Sounds good 👍 I'll look into it (since this is a Sonora integration feature, it seems like it belongs in this PR).

sri2k1us commented:
> > LGTM. However, when I ran this locally I found couple UX improvements that could be made:
> 
> Most of this UX is a direct import from the existing UX in the DE, and I was hoping UX changes for Sonora could have been brought up in #238, and to keep this PR just for page integration and URL / listing navigation UX.
> 
> FYI, if you don't want to checkout a PR and spin up Storybook locally, Chromium will still build our stories, even when our snapshot limit is reached: https://5ebad5c6b991d60022b1e72e-hutdrspubj.chromatic.com/?path=/story/metadata--metadata-view
> 
> With that said, I think I can implement some of these improvements in another PR, but let me address them 1 at a time, below:
> 
> > 1. Move the `Add` avu to toolbar and call it out in large screen instead of just icon. It took me few seconds to figure out where the add button was. Suggested change:
> 
> This is a good idea, however, the `+ AVU` header and the listing below it are reused for nested/child metadata listings:
> ![Edit AVU Dialog](https://user-images.githubusercontent.com/996408/102406378-b4c79880-3fa7-11eb-9558-93af04926c2b.png)
> 
> So I'm not sure where to move this button in the AVU Dialog 🤔
> 
> > 1. A `Back` button to take the user back to the listing.
> 
> This shouldn't be a problem and I can add this in this PR (this is the kind of Sonora UX navigation feedback I was looking for in this PR) 👍
> 
> > 1. I prefer the `Delete` button to be colored rather than its hover background color getting changed to red. In other places too, we set the color for icon (for example to `primary`) and not change the hover background color. Also in mobile, there is no hover rather the color shows up when you actually press the button.
> 
> I agree, the delete icon should probably be solid red, instead of red on hover, which will be nicer for mobile users.
> I'll save this for a UX improvement PR, though.
> 
> > 1. May be `Metadata` action could be linked into the file viewer too ? (This could be saved for later though)
> 
> Sounds good 👍 I'll look into it (since this is a Sonora integration feature, it seems like it belongs in this PR).

TBH - I did run and checked #238 locally but somehow didn't feel the same way about the `+` button when looking in SB at least. Here is the screenshot
![image](https://user-images.githubusercontent.com/1250790/102407659-6b347900-3fba-11eb-902d-518c204f4a45.png)

This looks like the whole thing is inside a dialog rather than a page along with tabs and others things. It looks different (perspective) when looked in Sonora page. But anyways, I try to keep this in mind.

psarando commented:
> This looks like the whole thing is inside a dialog rather than a page along with tabs and others things. It looks different (perspective) when looked in Sonora page. But anyways, I try to keep this in mind.

I see what you mean. Good point 👍

sri2k1us reviewed (commented):
LGTM so far. But It seems like I cannot view metadata for community data ? Is this planned for another PR ? This feature is available from DE right now.
![image](https://user-images.githubusercontent.com/1250790/102630365-76061f80-411a-11eb-9cd6-94baf48399f3.png)
![image](https://user-images.githubusercontent.com/1250790/102630487-9a61fc00-411a-11eb-9cee-df6d84594ae1.png)

sri2k1us commented:
> LGTM so far. But It seems like I cannot view metadata for community data ? Is this planned for another PR ? This feature is available from DE right now.
> ![image](https://user-images.githubusercontent.com/1250790/102630365-76061f80-411a-11eb-9cd6-94baf48399f3.png)
> ![image](https://user-images.githubusercontent.com/1250790/102630487-9a61fc00-411a-11eb-9cee-df6d84594ae1.png)

Oops I just realized I missed checking the row level dot menu. I found it there!. Anyway for consistency, `Metadata` Menu option should also be available from `Dot` menu in the toolbar.
![image](https://user-images.githubusercontent.com/1250790/102632970-cd59bf00-411d-11eb-90fb-13e61ea95c63.png)

psarando commented on src/components/utils/BackButton.js:
```diff
@@ -0,0 +1,34 @@
+
+import { ArrowBack } from "@material-ui/icons";
+
+export default function BackButton(props) {
+    const router = useRouter();
+    const theme = useTheme();
+    const isMobile = useMediaQuery(theme.breakpoints.down("xs"));
+    const { t } = useTranslation("common");
+
+    return (
+        <Button
+            color="primary"
+            variant={isMobile ? "text" : "contained"}
+            size="small"
+            startIcon={<ArrowBack fontSize="small" />}
+            onClick={() => router && router.back()}
+            {...props}
+        >
+            <Hidden xsDown>{t("back")}</Hidden>
+        </Button>
+    );
+}
```
I agree and I also prefer an icon button on mobile.
I'll make that change for this button before merging.

psarando commented:
> Oops I just realized I missed checking the row level dot menu. I found it there!. Anyway for consistency, `Metadata` Menu option should also be available from `Dot` menu in the toolbar.

Not sure why it's not showing up in the toolbar for read-only data, but I'll look into it 👍

psarando commented:
> 1. Move the `Add` avu to toolbar and call it out in large screen instead of just icon. It took me few seconds to figure out where the add button was. Suggested change:

> 3. I prefer the `Delete` button to be colored rather than its hover background color getting changed to red. In other places too, we set the color for icon (for example to `primary`) and not change the hover background color. Also in mobile, there is no hover rather the color shows up when you actually press the button.

I was going to save these updates for another UX improvement PR where I had intentions to add bigger changes for review (such as displaying child metadata in nested, collapsible rows, and only displaying form fields for editing AVUs in the edit dialog), but I already have some quick and easy fixes to address these issues ready for this PR 😅 

I added text to the "Add Metadata" button and just moved it to the far-right side of the listing title for now.
I also went ahead and removed the hover color of the AVU delete button and made its icon color red instead.
I also put the edit and delete buttons together in a button group (rather than the more hacky grid layout that was used before).
![Large screen UX improvements](https://user-images.githubusercontent.com/996408/102672461-132e7b80-414e-11eb-89b7-fb133401c886.png)

On mobile the "Add Metadata" button is still the `+` icon button:
![Mobile UX improvements](https://user-images.githubusercontent.com/996408/102672598-76201280-414e-11eb-873c-1640b2aa3979.png)

I rebased on master and I addressed all the other comments mentioned in this PR, so I'll go ahead and merge this PR with these UX fixes so they're available before the break, then I'll come back to the bigger metadata listing UX improvements later.

Thanks for the reviews! 👍 
Additional reviews on these changes still welcome (but may not be addressed until the New Year) 🙇
